ocs_ci/ocs/resources/cloud_manager.py

```diff
@@ -0,0 +1,281 @@
+from google.auth.exceptions import DefaultCredentialsError
+
+from ocs_ci.framework import config
+from ocs_ci.ocs import constants
+from ocs_ci.ocs.exceptions import TimeoutExpiredError
+from ocs_ci.utility import templating
+from ocs_ci.utility.utils import TimeoutSampler
+from tests.helpers import create_resource
+
+logger = logging.getLogger(name=__file__)
+
+
+class CloudManager(ABC):
+    """
+    Class containing all client types
+    """
+    aws_client, google_client, azure_client, s3comp_client = (None,) * 4
+
+    def __init__(self):
+        # TODO: solve credentials for clients (working with local creds for now)
+        self.aws_client = S3Client()
+        # TODO Need credentials to check
+        self.google_client = None
+        self.azure_client = None
+        self.s3comp_client = None
+
+
+class CloudClient(ABC):
+    """
+    Base abstract class for Cloud based API calls
+    """
+    client = None
+
+    def __init__(self, *args, **kwargs):
+        pass
+
+    def create_uls(self, name, region):
+        """
+        Super method that first logs the Underlying Storage creation and then calls
+        the appropriate implementation
+        """
+        logger.info(f"Creating Underlying Storage {name} in {region}")
+        self.internal_create_uls(name, region)
+
+    def delete_uls(self, name):
+        """
+        Super method that first logs the Underlying Storage deletion and then calls
+        the appropriate implementation
+        """
+        logger.info(f"Deleting ULS: {name}")
+        self.internal_delete_uls(name)
+
+    def get_all_uls_names(self):
+        pass
+
+    def verify_uls_exists(self, uls_name):
+        pass
+
+    @abstractmethod
+    def internal_create_uls(self, name, region):
+        pass
+
+    @abstractmethod
+    def internal_delete_uls(self, name):
+        pass
+
+
+class S3Client(CloudClient):
+    """
+    Implementation of a S3 Client using the S3 API
+    """
+
+    def __init__(self, key_id=None, access_key=None, endpoint="https://s3.amazonaws.com",
+                 verify=True, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        if key_id and access_key:
+            self.client = boto3.resource(
+                's3', verify=verify, endpoint_url=endpoint,
+                aws_access_key_id=key_id,
+                aws_secret_access_key=access_key
+            )
+            self.access_key = key_id
+            self.secret_key = access_key
+        else:
+            self.client = boto3.resource('s3', endpoint_url=endpoint)
+            # create a secret for the underlying storage to use
+            session = boto3.Session()
+            credentials = session.get_credentials()
+            credentials = credentials.get_frozen_credentials()
+            self.access_key = credentials.access_key
+            self.secret_key = credentials.secret_key
+        bs_secret_data = templating.load_yaml(constants.MCG_BACKINGSTORE_SECRET_YAML)
```

Comment:
Suggestion: This can be defined outside of this class as a new class `AWSBackingstoreSecret` that can be used by others.

Reply:
This suggestion is a nice idea but since this pr needs to be pushed to allow further work in other areas I've opened an issue with a link to this pr so we can implement it in a different time:
https://github.com/red-hat-storage/ocs-ci/issues/1854